File: bench/trees.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

func arg(at, fallback int) int {
	if len(os.Args) <= at+1 {
		return fallback
	}
	n, err := strconv.Atoi(os.Args[at+1])
	if err != nil {
		return fallback
	}
	return n
}

type Node struct {
	Left, Right *Node
	Value       int
}

func build(depth, seed int) *Node {
	n := &Node{Value: depth + seed}
	if depth > 0 {
		n.Left = build(depth-1, seed)
		n.Right = build(depth-1, seed)
	}
	return n
}

func count(n *Node) int {
	total := 1 + n.Value
	if n.Left != nil {
		total += count(n.Left)
	}
	if n.Right != nil {
		total += count(n.Right)
	}
	return total
}

func main() {
	maxDepth, seed := arg(0, 14), arg(1, 1)
	longLived := build(maxDepth, seed)
	total := 0

	for d := 4; d <= maxDepth; d += 2 {
		iters := 1
		for k := maxDepth - d + 4; k > 0; k-- {
			iters *= 2
		}
		for i := 0; i < iters; i++ {
			total += count(build(d, seed))
		}
	}

	fmt.Printf("trees %d long %d\n", total, count(longLived))
}

File: bench/trees.js
// bun trees.js — mirror of trees.b / trees.go (binary-trees)
function build(depth, seed) {
    const n = { left: null, right: null, value: depth + seed };
    if (depth > 0) {
        n.left = build(depth - 1, seed);
        n.right = build(depth - 1, seed);
    }
    return n;
}

function count(n) {
    let total = 1 + n.value;
    if (n.left !== null) total += count(n.left);
    if (n.right !== null) total += count(n.right);
    return total;
}

const maxDepth = Number(process.argv[2] ?? 14);
const seed = Number(process.argv[3] ?? 1);
const longLived = build(maxDepth, seed);
let total = 0;

for (let d = 4; d <= maxDepth; d += 2) {
    let iters = 1;
    for (let k = maxDepth - d + 4; k > 0; k--) iters *= 2;
    for (let i = 0; i < iters; i++) total += count(build(d, seed));
}

console.log(`trees ${total} long ${count(longLived)}`);
